### ISSUE TITLE
System.Runtime.InteropServices.MarshalDirectiveException: 'Method's type signature is not PInvoke compatible.'

### ISSUE THREAD

**Issue by futureflsl** [SciSharp/LLamaSharp#484]
I use LLamaSharp 0.9.1,and can not run well,it error occured
var model = LLamaWeights.LoadFromFile(parameters);  results in an error at runtime:
System.Runtime.InteropServices.MarshalDirectiveException: 'Method's type signature is not PInvoke compatible.'
I found issue https://github.com/SciSharp/LLamaSharp/issues/14,this problem is like me,but no solution.my environment is:
vs2019
netframework4.8.0
LLamaSharp 0.9.1
LLamaSharp.Backend.Cpu ver0.9.1)

**Comment by SINAPSA-IC** [comment]
2024-Feb-03:
I've encountered the same issue:
System.Runtime.InteropServices.MarshalDirectiveException: 'Method's type signature is not PInvoke compatible.'
The exception occurs in LoadFromFile(...)
So, the exe stops dead in the first steps.
I hope this is fixable. Thank you.

Variations leading to the same error:
- built the solution as Debug
- built the solution as Release
- Project menu -> [project] Properties 
- -> Build -> 
- checked/unchecked: "Prefer 32-bit", "Allow unsafe code";
- selected from Platform target: "Any CPU", "x86" (useless), "x64"

System:
x64 Intel
Visual Studio 2022 CE
.NET 4.8
LLamaSharp 0.9.1
LLamaSharp.Backend.Cpu 0.9.1

**Comment by martindevans** [comment]
I think that means that one of the native llama.cpp methods which LLamaSharp is not quite compatible with the .NET Framework pinvoke system. That can probably be worked around, if either of you can track down _exactly_ which method this is I can take a look at it.

To track it down just stick down a breakpoint on the LoadFromFile method and step _into_ it, eventually you'll get to a method which crashes when called. That method is the one that'll need to be modified to work around this.

**Comment by SINAPSA-IC** [comment]
Thank you for responding.

Will do.

Could it be that certain processors handle that method fine, while others don't?.. This _could explain why _some users only signal this issue, coulda, woulda, shoulda.

**Comment by martindevans** [comment]
I think anyone running LLamaSharp on .NET Framework would have this issue. But framework is pretty rare to use these days.

**Comment by SINAPSA-IC** [comment]
I found the culprit, so it seems.
Hope this helps. A bit.

The way to crash'n burn was:

**1 of 2)**
the line:
**using (@params.ToLlamaModelParams(out result))**

in method:
public static LLamaWeights LoadFromFile(IModelParams @params)
{
LLamaModelParams result;
**using (@params.ToLlamaModelParams(out result))**
{
SafeLlamaModelHandle safeLlamaModelHandle = SafeLlamaModelHandle.LoadFromFile(@params.ModelPath, result);
foreach (LoraAdapter loraAdapter in @params.LoraAdapters)
{
if (!string.IsNullOrEmpty(loraAdapter.Path) && !(loraAdapter.Scale <= 0f))
{
safeLlamaModelHandle.ApplyLoraFromFile(loraAdapter.Path, loraAdapter.Scale, @params.LoraBase);
}
}
return new LLamaWeights(safeLlamaModelHandle);
}
}


**2 of 2)**
That line shouts to public static class IModelParamsExtensions
where:

something happened in the line:
result = NativeApi.llama_model_default_params();

in method (well, here's the whole litany):
public unsafe static IDisposable ToLlamaModelParams(this IModelParams @params, out LLamaModelParams result)
{
//IL_0089: Unknown result type (might be due to invalid IL or missing references)
//IL_008e: Unknown result type (might be due to invalid IL or missing references)
//IL_0093: Unknown result type (might be due to invalid IL or missing references)
//IL_00e0: Unknown result type (might be due to invalid IL or missing references)
//IL_00e5: Unknown result type (might be due to invalid IL or missing references)
//IL_00ea: Unknown result type (might be due to invalid IL or missing references)
if (@params.UseMemoryLock && !NativeApi.llama_mlock_supported())
{
throw new NotSupportedException("'UseMemoryLock' is not supported (llama_mlock_supported() == false)");
}
if (@params.UseMemorymap && !NativeApi.llama_mmap_supported())
{
throw new NotSupportedException("'UseMemorymap' is not supported (llama_mmap_supported() == false)");
}
GroupDisposable groupDisposable = new GroupDisposable();
**result = NativeApi.llama_model_default_params();**
result.main_gpu = @params.MainGpu;
result.n_gpu_layers = @params.GpuLayerCount;
result.use_mlock = @params.UseMemoryLock;
result.use_mmap = @params.UseMemorymap;
result.vocab_only = @params.VocabOnly;
MemoryHandle val = groupDisposable.Add(@params.TensorSplits.Pin());
result.tensor_split = (float*)((MemoryHandle)(ref val)).Pointer;
if (@params.MetadataOverrides.Count == 0)
{
result.kv_overrides = (LLamaModelMetadataOverride*)(void*)IntPtr.Zero;
}
else
{
LLamaModelMetadataOverride[] array = new LLamaModelMetadataOverride[@params.MetadataOverrides.Count + 1];
val = groupDisposable.Add(MemoryExtensions.AsMemory<LLamaModelMetadataOverride>(array).Pin());
result.kv_overrides = (LLamaModelMetadataOverride*)((MemoryHandle)(ref val)).Pointer;
for (int i = 0; i < @params.MetadataOverrides.Count; i++)
{
MetadataOverride metadataOverride = @params.MetadataOverrides[i];
LLamaModelMetadataOverride lLamaModelMetadataOverride = default(LLamaModelMetadataOverride);
lLamaModelMetadataOverride.Tag = metadataOverride.Type;
LLamaModelMetadataOverride dest = lLamaModelMetadataOverride;
metadataOverride.WriteValue(ref dest);
System.ReadOnlySpan<char> chars = MemoryExtensions.AsSpan(metadataOverride.Key);
EncodingExtensions.GetBytes(output: new System.Span<byte>((void*)dest.key, 128), encoding: Encoding.UTF8, chars: chars);
array[i] = dest;
}
}
return groupDisposable;
}

**Comment by martindevans** [comment]
So you think it's probably the `llama_model_default_params` method that's the issue?

**Comment by SINAPSA-IC** [comment]
Yes, looks like it, and I stop at this conclusion, since I cannot-I do not know how to change anything.

**Comment by martindevans** [comment]
Thanks for putting in the effort to narrow it down :)

**Comment by SINAPSA-IC** [comment]
No problem :)
Happy! to help. My knowledge of C++ stops at VS6.
Anyway, I may be wrong, but as I see it...
result = NativeApi.llama_model_default_params();
provides a default value, to be used before parameter micromanagement below it
Perhaps it should be moved in a catch, to be used if the micromanagement goes wrong in the try?
Shot in the dark...

**Comment by SINAPSA-IC** [comment]
Hi there,
I don't know why this issue does not get the attention that seemingly deserves, since I speaking for myself cannot even start a program by loading a model, using any of these, in all possible combinations:
- .NET Frameworks 2.0, 4.8, 6.0, 8.0
- llamasharp 0.9 and 0.10
- Console Apps and WinForms Apps
The code simply stops at the line where a model is being loaded (not) with the forwarded parameters, in the method NativeApi.llama_model_default_params(), and as such Nothing can be done from that point onwards.
Please, someone provide some info about this Critical issue. It's been almost one month since signaling it. Thank you.

**Comment by SINAPSA-IC** [comment]
I've been trying with 0.10 in the hope that it works; alas, it is not the case. 

VS2022C, Console App, Target .NET 8.0.

The code is simple, made even more simple with a single parameter, yet the problem is the same.

	var parameters = new ModelParams(modelPath)
	{
		ContextSize whatever
	};

This ^ calls upon:

	[JsonConstructor]
	public ModelParams(string modelPath)
	{
		ModelPath = modelPath;
	}

which in turn calls:

	public TensorSplitsCollection()
	{
	}

which FWIW ends the .exe process with code -1073741795.

... which in 0.10 is different from 0.9, since it does _not get to Native API's default_params anymore.

Looks like the exit point depends on the version of .NET, since with 4.8 it gets to the default_params method in NativeAPI, whereas with 8.0 stops at the entry point in the constructor of  TensorSplitsCollection.

**Comment by martindevans** [comment]
> I don't know why this issue does not get the attention that it seemingly deserves

Any issue only gets attention because a motivated developer contributes their spare time and expertise to work on it! No one is being paid to work on LLamaSharp.

As I understood it, this issue was about .NET Framework. However, from what you've said in your most recent message it looks like it's also not working for you with .NET 6/7/8. Is that correct?

**Comment by SINAPSA-IC** [comment]
Yes indeed - it does not work with any of these: 2.0, 4.8, 6.0, 7.0, 8.0.

The fact that others do not encounter this problem makes it even more frustrating; stopping at step 1 - when loading a model, before anything else, yet not everywhere.

**Comment by martindevans** [comment]
Are you _absolutely certain_ that you correctly configured the dotnet version? I've had a look through the docs and I can't see how this error could happen for you on NET8, but not for others.

**Comment by SINAPSA-IC** [comment]
I think so, given that I've followed the instructions for installing the 0.10.0 packages of llamasharp and backend.cpu -  anyway, I have no idea how/what to do otherwise.

Models tested on:
- downloaded from: https://gpt4all.io/index.html
- - mistral-7b-instruct-v0.1.Q4_0.gguf
- - gpt4all-falcon-newbpe-q4_0.gguf

The entire Console App / .NET 8.0 is the example below taken from: 
https://blog.chuckbeasley.com/321/
only modified with GpuLayerCount=4 instead of 35.

The program starts, displays a log, then when sending the parameters it crashes in: 
```
public TensorSplitsCollection()
{ //here
}
```


```
using LLama.Common;
using LLama;
using LLama.Native;

namespace ConsoleApp1
{
	internal class Program
	{


		static async Task Main(string[ ] args)
		{
			Program p = new Program();
			string smodelpath = "mistral-7b-instruct-v0.1.Q4_0.gguf";
			var prompt = "Transcript of a dialog, where the User interacts with an Assistant named Bob. Bob is helpful, kind, honest, good at writing, and never fails to answer the User's requests immediately and with precision.\r\n\r\nUser: Hello, Bob.\r\nBob: Hello. How may I help you today?\r\nUser: Please tell me the largest city in Europe.\r\nBob: Sure. The largest city in Europe is Moscow, the capital of Russia.\r\nUser:"; // use the "chat-with-bob" prompt here.
			NativeLibraryConfig.Instance.WithLogs();

			// Load a model

			var parameters = new ModelParams(smodelpath)
			{
				ContextSize = 4096,
				GpuLayerCount = 4,
				UseMemoryLock = true,
				UseMemorymap = true
			};
			using var model = LLamaWeights.LoadFromFile(parameters);
			// Initialize a chat session
			using var context = model.CreateContext(parameters);
			var ex = new InteractiveExecutor(context);
			ChatSession session = new ChatSession(ex);

			// show the prompt
			Console.WriteLine();
			Console.Write(prompt);

			// run the inference in a loop to chat with LLM
			while (prompt != "stop")
			{
				await foreach (
					var text in session.ChatAsync(
						new ChatHistory.Message(AuthorRole.User, prompt),
						new InferenceParams
						{
							Temperature = 0.1f,
							AntiPrompts = new List<string> { "User:" }
						}))
				{
					Console.Write(text);
				}
				prompt = Console.ReadLine() ?? "";
			}
			//
			//
			//// save the session
			//session.SaveSession("SavedSessionPath");
		}

	}   //end of class Program
}	//end of namespace ConsoleApp1

```

**Comment by SINAPSA-IC** [comment]
OK now...

It turns out that it was a question of RAM, if not of processor; I suspect that the RAM is the reason:
- the problem starts with the model path
- which means the model file
- which /must be loaded into RAM
- where it should fit

so:

I've tried the Apps and the code above on these low-end systems:
- Intel E7500, 4 GB (yes) RAM:
-- the Apps crash at the instructions signaled in this thread: PInvoke for WinForms Apps and TensorSplits for Console Apps
--- this I should've linked to the fact that GPT4All doesn't _load said models either, but nope...
-- Log says: NO AVX

- Intel i5 Core 8265U, 8 GB RAM:
-- the Console App on .NET 8.0 runs OK
--- like GPT4All does
-- Log says: AVX2

The next one to use/test is an (i5 + 12 GB RAM), but I presume it's clear...
Haven't tested a WinForms on the i5, but - likewise:

**Not enough RAM to load a model in = App crash.**

I guess some technical writing about this simple situation is in order... like, basic requirements... my apologies if such info does exist.

I for myself got mindlessly caught in this all too visible trap, and have gotten there you too, so please accept my sincere apologies! I cannot give you back the time spent on this. Time is the fire in which we burn...

So this should be closed and possibly documented somewhere, red font in Capitals, like...

Thank you! for your Time! and interest in this issue.

**Comment by martindevans** [comment]
Aha, good job narrowing that down!
